<?xml version="1.0" encoding="utf-8"?><testsuites name="pytest tests"><testsuite name="pytest" errors="0" failures="1" skipped="0" tests="27" time="2.524" timestamp="2025-07-01T20:06:16.228842+10:00" hostname="DESKTOP-0Q8HRTN"><testcase classname="tests.test_favourites" name="test_load_favourites_file_not_found" file="tests\test_favourites.py" line="10" time="0.002"><system-out>--------------------------------- Captured Log ---------------------------------

--------------------------------- Captured Out ---------------------------------

</system-out><system-err>--------------------------------- Captured Err ---------------------------------

</system-err></testcase><testcase classname="tests.test_favourites" name="test_load_favourites_json_error" file="tests\test_favourites.py" line="30" time="0.005"><system-out>--------------------------------- Captured Log ---------------------------------

--------------------------------- Captured Out ---------------------------------

</system-out><system-err>--------------------------------- Captured Err ---------------------------------

</system-err></testcase><testcase classname="tests.test_favourites" name="test_load_favourites_success" file="tests\test_favourites.py" line="20" time="0.004"><system-out>--------------------------------- Captured Log ---------------------------------

--------------------------------- Captured Out ---------------------------------

</system-out><system-err>--------------------------------- Captured Err ---------------------------------

</system-err></testcase><testcase classname="tests.test_favourites" name="test_save_favourite_add_new_city" file="tests\test_favourites.py" line="42" time="0.004"><system-out>--------------------------------- Captured Log ---------------------------------

--------------------------------- Captured Out ---------------------------------

</system-out><system-err>--------------------------------- Captured Err ---------------------------------

</system-err></testcase><testcase classname="tests.test_favourites" name="test_save_favourite_empty_city_string" file="tests\test_favourites.py" line="62" time="0.001"><system-out>--------------------------------- Captured Log ---------------------------------

--------------------------------- Captured Out ---------------------------------

</system-out><system-err>--------------------------------- Captured Err ---------------------------------

</system-err></testcase><testcase classname="tests.test_favourites" name="test_save_favourite_non_string_input" file="tests\test_favourites.py" line="72" time="0.001"><system-out>--------------------------------- Captured Log ---------------------------------

--------------------------------- Captured Out ---------------------------------

</system-out><system-err>--------------------------------- Captured Err ---------------------------------

</system-err></testcase><testcase classname="tests.test_favourites" name="test_save_favourite_remove_existing_city" file="tests\test_favourites.py" line="52" time="0.006"><system-out>--------------------------------- Captured Log ---------------------------------

--------------------------------- Captured Out ---------------------------------

</system-out><system-err>--------------------------------- Captured Err ---------------------------------

</system-err></testcase><testcase classname="tests.test_main" name="test_on_close_calls_destroy_and_logs" file="tests\test_main.py" line="24" time="0.003"><system-out>--------------------------------- Captured Log ---------------------------------

--------------------------------- Captured Out ---------------------------------

</system-out><system-err>--------------------------------- Captured Err ---------------------------------

</system-err></testcase><testcase classname="tests.test_main" name="test_on_select_favourite_triggers_search_and_update" file="tests\test_main.py" line="8" time="0.003"><system-out>--------------------------------- Captured Log ---------------------------------

--------------------------------- Captured Out ---------------------------------

</system-out><system-err>--------------------------------- Captured Err ---------------------------------

</system-err></testcase><testcase classname="tests.test_ui_components" name="test_handle_search_location_not_found" file="tests\test_ui_components.py" line="89" time="0.146"><system-out>--------------------------------- Captured Log ---------------------------------
ERROR    WeatherViewLogger:ui_components.py:150 Could not update UI: KeyError - 'condition'
--------------------------------- Captured Out ---------------------------------

</system-out><system-err>--------------------------------- Captured Err ---------------------------------

</system-err></testcase><testcase classname="tests.test_ui_components" name="test_handle_search_success" file="tests\test_ui_components.py" line="48" time="0.117"><failure message="AssertionError: Expected 'set_dynamic_background' to have been called once. Called 2 times.&#10;Calls: [call(&lt;tkinter.Tk object .&gt;, 'Clouds'),&#10; call().__iter__(),&#10; call(&lt;tkinter.Tk object .&gt;, 'Clouds')].">mock_reset_refresh = &lt;MagicMock name='reset_auto_refresh' id='1650396967472'&gt;
mock_update_fav = &lt;MagicMock name='update_fav_button' id='1650396957056'&gt;
mock_embed = &lt;MagicMock name='embed_chart' id='1650396957392'&gt;
mock_create_fig = &lt;MagicMock name='create_forecast_figure' id='1650396968816'&gt;
mock_load_icon = &lt;MagicMock name='load_weather_icon' id='1650396968480'&gt;
mock_set_bg = &lt;MagicMock name='set_dynamic_background' id='1650396968144'&gt;
mock_get_detailed = &lt;MagicMock name='get_detailed_forecast_by_city' id='1650396967808'&gt;
mock_get_forecast = &lt;MagicMock name='get_forecast_by_city' id='1650396967136'&gt;
mock_get_weather = &lt;MagicMock name='get_weather_by_city' id='1650396966800'&gt;
mock_ui = {'chart_frame': &lt;MagicMock id='1650396970160'&gt;, 'city_label': &lt;MagicMock id='1650396964112'&gt;, 'condition_label': &lt;MagicMock id='1650396964448'&gt;, 'forecast_cards_container': &lt;MagicMock id='1650396970496'&gt;, ...}
mock_unit_var = &lt;MagicMock id='1650396966128'&gt;

    @patch('ui_components.get_weather_by_city')
    @patch('ui_components.get_forecast_by_city')
    @patch('ui_components.get_detailed_forecast_by_city')
    @patch('ui_components.set_dynamic_background')
    @patch('ui_components.load_weather_icon')
    @patch('ui_components.create_forecast_figure')
    @patch('ui_components.embed_chart')
    @patch('ui_components.update_fav_button')
    @patch('ui_components.reset_auto_refresh')
    def test_handle_search_success(
        mock_reset_refresh, mock_update_fav, mock_embed, mock_create_fig,
        mock_load_icon, mock_set_bg, mock_get_detailed, mock_get_forecast,
        mock_get_weather, mock_ui, mock_unit_var
    ):
        """
        Tests the successful path of handle_search, where all API calls return valid data.
        """
        # Mock API responses
        mock_get_weather.return_value = {
            "city": "London", "temperature": 15, "condition": "Clouds",
            "humidity": 80, "wind_speed": 5, "day": "Tuesday",
            "date": "2025-07-01", "time": "12:00", "icon": "04d"
        }
        mock_get_forecast.return_value = []
        mock_get_detailed.return_value = []
        mock_load_icon.return_value = "fake_photo_image"
    
        # Action
        ui_components.handle_search(mock_ui, mock_unit_var)
    
        # Assertions
        mock_ui["status_label"].config.assert_any_call(text="Loading...", foreground="black")
        mock_get_weather.assert_called_once_with("London", "metric")
        mock_ui["city_label"].config.assert_called_with(text="London", foreground="black")
        mock_ui["temp_label"].config.assert_called_with(text="Temperature: 15Â°C")
&gt;       mock_set_bg.assert_called_once()

tests\test_ui_components.py:84: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = &lt;MagicMock name='set_dynamic_background' id='1650396968144'&gt;

    def assert_called_once(self):
        """assert that the mock was called only once.
        """
        if not self.call_count == 1:
            msg = ("Expected '%s' to have been called once. Called %s times.%s"
                   % (self._mock_name or 'mock',
                      self.call_count,
                      self._calls_repr()))
&gt;           raise AssertionError(msg)
E           AssertionError: Expected 'set_dynamic_background' to have been called once. Called 2 times.
E           Calls: [call(&lt;tkinter.Tk object .&gt;, 'Clouds'),
E            call().__iter__(),
E            call(&lt;tkinter.Tk object .&gt;, 'Clouds')].

..\..\..\..\AppData\Local\Programs\Python\Python313\Lib\unittest\mock.py:958: AssertionError</failure><system-out>--------------------------------- Captured Log ---------------------------------
ERROR    WeatherViewLogger:ui_components.py:62 Failed to retrieve Current Weather from API
--------------------------------- Captured Out ---------------------------------

</system-out><system-err>--------------------------------- Captured Err ---------------------------------

</system-err></testcase><testcase classname="tests.test_ui_components" name="test_load_weather_icon_fallback_to_default" file="tests\test_ui_components.py" line="131" time="0.003"><system-out>--------------------------------- Captured Log ---------------------------------

--------------------------------- Captured Out ---------------------------------

</system-out><system-err>--------------------------------- Captured Err ---------------------------------

</system-err></testcase><testcase classname="tests.test_ui_components" name="test_load_weather_icon_success" file="tests\test_ui_components.py" line="114" time="0.003"><system-out>--------------------------------- Captured Log ---------------------------------

--------------------------------- Captured Out ---------------------------------

</system-out><system-err>--------------------------------- Captured Err ---------------------------------

</system-err></testcase><testcase classname="tests.test_utils" name="test_create_forecast_figure" file="tests\test_utils.py" line="91" time="0.096"><system-out>--------------------------------- Captured Log ---------------------------------

--------------------------------- Captured Out ---------------------------------

</system-out><system-err>--------------------------------- Captured Err ---------------------------------

</system-err></testcase><testcase classname="tests.test_utils" name="test_get_user_city_failure" file="tests\test_utils.py" line="126" time="0.002"><system-out>--------------------------------- Captured Log ---------------------------------

--------------------------------- Captured Out ---------------------------------

</system-out><system-err>--------------------------------- Captured Err ---------------------------------

</system-err></testcase><testcase classname="tests.test_utils" name="test_get_user_city_success" file="tests\test_utils.py" line="110" time="0.003"><system-out>--------------------------------- Captured Log ---------------------------------

--------------------------------- Captured Out ---------------------------------

</system-out><system-err>--------------------------------- Captured Err ---------------------------------

</system-err></testcase><testcase classname="tests.test_utils" name="test_on_remove_favourite" file="tests\test_utils.py" line="53" time="0.003"><system-out>--------------------------------- Captured Log ---------------------------------

--------------------------------- Captured Out ---------------------------------

</system-out><system-err>--------------------------------- Captured Err ---------------------------------

</system-err></testcase><testcase classname="tests.test_utils" name="test_on_save_favourite_new_city" file="tests\test_utils.py" line="27" time="0.003"><system-out>--------------------------------- Captured Log ---------------------------------

--------------------------------- Captured Out ---------------------------------

</system-out><system-err>--------------------------------- Captured Err ---------------------------------

</system-err></testcase><testcase classname="tests.test_utils" name="test_set_dynamic_background[clear]" file="tests\test_utils.py" line="70" time="0.003"><system-out>--------------------------------- Captured Log ---------------------------------

--------------------------------- Captured Out ---------------------------------

</system-out><system-err>--------------------------------- Captured Err ---------------------------------

</system-err></testcase><testcase classname="tests.test_utils" name="test_set_dynamic_background[drizzle]" file="tests\test_utils.py" line="70" time="0.005"><system-out>--------------------------------- Captured Log ---------------------------------

--------------------------------- Captured Out ---------------------------------

</system-out><system-err>--------------------------------- Captured Err ---------------------------------

</system-err></testcase><testcase classname="tests.test_utils" name="test_set_dynamic_background[few clouds]" file="tests\test_utils.py" line="70" time="0.003"><system-out>--------------------------------- Captured Log ---------------------------------

--------------------------------- Captured Out ---------------------------------

</system-out><system-err>--------------------------------- Captured Err ---------------------------------

</system-err></testcase><testcase classname="tests.test_utils" name="test_set_dynamic_background[light snow]" file="tests\test_utils.py" line="70" time="0.003"><system-out>--------------------------------- Captured Log ---------------------------------

--------------------------------- Captured Out ---------------------------------

</system-out><system-err>--------------------------------- Captured Err ---------------------------------

</system-err></testcase><testcase classname="tests.test_utils" name="test_set_dynamic_background[mist]" file="tests\test_utils.py" line="70" time="0.003"><system-out>--------------------------------- Captured Log ---------------------------------

--------------------------------- Captured Out ---------------------------------

</system-out><system-err>--------------------------------- Captured Err ---------------------------------

</system-err></testcase><testcase classname="tests.test_utils" name="test_set_dynamic_background[shower rain]" file="tests\test_utils.py" line="70" time="0.005"><system-out>--------------------------------- Captured Log ---------------------------------

--------------------------------- Captured Out ---------------------------------

</system-out><system-err>--------------------------------- Captured Err ---------------------------------

</system-err></testcase><testcase classname="tests.test_utils" name="test_set_dynamic_background[thunderstorms]" file="tests\test_utils.py" line="70" time="0.009"><system-out>--------------------------------- Captured Log ---------------------------------

--------------------------------- Captured Out ---------------------------------

</system-out><system-err>--------------------------------- Captured Err ---------------------------------

</system-err></testcase><testcase classname="tests.test_utils" name="test_set_dynamic_background[unknown condition]" file="tests\test_utils.py" line="70" time="0.005"><system-out>--------------------------------- Captured Log ---------------------------------

--------------------------------- Captured Out ---------------------------------

</system-out><system-err>--------------------------------- Captured Err ---------------------------------

</system-err></testcase><testcase classname="tests.test_utils" name="test_update_fav_button_city_is_favourite" file="tests\test_utils.py" line="42" time="0.003"><system-out>--------------------------------- Captured Log ---------------------------------

--------------------------------- Captured Out ---------------------------------

</system-out><system-err>--------------------------------- Captured Err ---------------------------------

</system-err></testcase></testsuite></testsuites>